data/events/alerts.go

```diff
@@ -0,0 +1,381 @@
+	case strings.HasSuffix(msg.Topic, ".data.deviceData.alerts"):
+		return c.consumeDeviceData(ctx, session, msg)
+	default:
+		c.Logger.WithField("topic", msg.Topic).Infof("UNEXPECTED: topic; ignoring")
+	}
+
+	return nil
+}
+
+func (c *Consumer) consumeAlertsConfigs(ctx context.Context,
+	session sarama.ConsumerGroupSession, msg *sarama.ConsumerMessage) error {
+
+	cfg := &alerts.Config{}
+	if err := unmarshalMessageValue(msg.Value, cfg); err != nil {
+		return err
+	}
+	lgr := log.LoggerFromContext(ctx)
+	lgr.WithField("cfg", cfg).Info("consuming an alerts config message")
+
+	err := c.evaluateAlerts(ctx, cfg.FollowedUserID, cfg.UserID)
+	if err != nil {
+		return errors.Wrap(err, "Unable to evalaute alerts configs triggered event")
+	}
+	session.MarkMessage(msg, "")
+	lgr.WithField("message", msg).Debug("marked")
+	return nil
+}
+
+func (c *Consumer) consumeDeviceData(ctx context.Context,
+	session sarama.ConsumerGroupSession, msg *sarama.ConsumerMessage) error {
+
+	blood := &glucose.Glucose{}
+	if err := unmarshalMessageValue(msg.Value, blood); err != nil {
+		return err
+	}
+	lgr := log.LoggerFromContext(ctx)
+	lgr.WithField("data", blood).Info("consuming a device data message")
+
+	if blood.UserID == nil {
+		return errors.New("Unable to retrieve alerts configs: userID is nil")
+	}
+	followerConfigs, err := c.Alerts.List(ctx, *blood.UserID)
+	if err != nil {
+		return err
+	}
+	serverSessionToken, err := c.Tokens.ServerSessionToken()
+	if err != nil {
+		return err
+	}
+	ctxWithAuth := auth.NewContextWithServerSessionToken(ctx, serverSessionToken)
+	permittedConfigs := []*alerts.Config{}
+	for _, followerAlert := range followerConfigs {
+		perms, err := c.Permissions.GetUserPermissions(ctxWithAuth, followerAlert.UserID, *blood.UserID)
+		if err != nil {
+			lgr.WithError(err).Error("Unable to retrieve user permissions")
+			continue
+		}
+		if _, ok := perms[permission.Follow]; ok {
+			permittedConfigs = append(permittedConfigs, followerAlert)
+		}
+	}
+	for _, permittedConfig := range permittedConfigs {
+		err = c.evaluateAlerts(ctx, permittedConfig.FollowedUserID, permittedConfig.UserID)
+		if err != nil {
+			lgr.WithError(err).Error("Unable to evaluate device data triggered event")
+			continue
+		}
+	}
+
+	session.MarkMessage(msg, "")
+	lgr.WithField("message", msg).Debug("marked")
+	return nil
+}
+
+// ensureContextHasLogger is a convenience method that makes it easier for
+// methods to find a logger.
+func (c *Consumer) ensureContextHasLogger(ctx context.Context) (context.Context, error) {
+	if ctxLgr := log.LoggerFromContext(ctx); ctxLgr != nil {
+		return ctx, nil
+	}
+	if c.Logger != nil {
+		return log.NewContextWithLogger(ctx, c.Logger), nil
+	}
+	lgr, err := logjson.NewWithDefaults()
+	if err != nil {
+		return ctx, err
+	}
+	return log.NewContextWithLogger(ctx, lgr), nil
+}
+
+func (c *Consumer) evaluateAlerts(ctx context.Context, followedUserID, userID string) (err error) {
```

Comment:
Both kafka topics must be keyed by the followed user id, to avoid race conditions by ensuring that alerts for the same followed user are evaluated by the same consumer.